### ISSUE TITLE
Backup/Restart Not working with Latest Truffle/Solidity on RHEL 7.6 nor on Ubuntu 16 or 18

### ISSUE THREAD

**Issue by ghassanius** [ethereum/go-ethereum#19263]
#### System information

Geth version: `geth version`
Geth
Version: 1.9.0-unstable
Git Commit: 3eff652a7b606f25d43bef6ccb998b8e306f8a75
Architecture: amd64
Protocol Versions: [63 62]
Network Id: 1
Go Version: go1.11.5
Operating System: linux
GOPATH=
GOROOT=/usr/local/go

OS & Version: Windows/Linux/OSX
   Static hostname: MonimoveRedhat
         Icon name: computer-vm
           Chassis: vm
        Machine ID: 6ca77495007bXXXXXXX8f0e34d11
           Boot ID: 12699d2ae8034XXXXXX95ab3a326c
    Virtualization: microsoft
  Operating System: Red Hat Enterprise Linux Server 7.6 (Maipo)
       CPE OS Name: cpe:/o:redhat:enterprise_linux:7.6:GA:server
            Kernel: Linux 3.10.0-957.5.1.el7.x86_64
      Architecture: x86-64

Truffle Version:
Truffle v5.0.8 (core: 5.0.8)
Solidity v0.5.0 (solc-js)
Node v11.11.0
Web3.js v1.0.0-beta.37

Commit hash : (if `develop`)

#### Expected behaviour
1- The data should be backedup and I should be able to find the information I previously added whenever I import/restore the old backup.
![image](https://user-images.githubusercontent.com/25767691/54363249-3bcd3b80-4673-11e9-88f7-863606adfcbb.png)

2- The getPastLog functions is working fine on the first day, and it should continue working fine in the next days.

#### Actual behaviour
1- The data is gone and no trace for any information previously added.
![image](https://user-images.githubusercontent.com/25767691/54363269-48519400-4673-11e9-86f3-1f51962bcf2e.png)

2- After a while (1 day) the getPastLogs function stops working and displays the below issue (fig.2) while all other APIs are working fine (happening only on RHEL 7.6).

#### Steps to reproduce the behaviour for Issue 1
I tried using:
- admin.importChain(file) & admin.exportChain(file).
- copy/pasting the whole directory of myDataDir

All of this used to work fine on older versions of web3, solidity and truffle. But to use the getPastLog function we needed to upgrade to the newest web3 which only works with the new solidity and truffle version. 

My environment is a NodeJS application that communicates with the Blockchain and save information on it. When I save some through APIs, I'm able to see the data using Get requests. When I kill the process of geth and then relaunch it using "geth --datadir ./myDataDir --rpc --targetgaslimit 18446744073709551615 --networkid 1441 console 2>> myEth.log" I get a complete new blockchain as if it's new! Here are the full steps:

```
ps -aux | grep geth
sudo kill -9 xxxxx
pm2 stop NodeApp
pm2 delete NodeApp
cd ProjectDirectory
cd Blockchain
tmux
geth --datadir ./myDataDir --rpc --targetgaslimit 18446744073709551615 --networkid 1441 console 2>> myEth.log
miner.start()
personal.unlockAccount(personal.listAccounts[0], "XxXXxXXXXXxxXXxXXXxXX", 0)
ctrl b+d (to keep the session running in background)

cd ../Solidity
rm -r build/
rm ../Nodejs/server/helpers/address.json
mv address.json address.json_backup

vi truffle.js (update the account address)
truffle deploy --network MyNetwork
cp address.json ../Nodejs/server/helpers/address.json
rm ../Nodejs/server/contracts/*
cp -p build/contracts/* ../Nodejs/server/contracts/
cd ../Nodejs/server
pm2 start server.js --log-date-format 'DD-MM HH:mm:ss.SSS'
```

#### Backtrace
UBUNTU TESTING

If you don’t delete /build and didn’t replace address.json and used the same blockchain datadir

truffle deploy --network TestingNetwork

![image](https://user-images.githubusercontent.com/25767691/54423718-fdd72280-4719-11e9-889d-e05d51bacb62.png)

If you delete /build and didnt replace address.json and used the same blockchain datadir
rm -r build/
truffle deploy --network TestingNetwork
rm ../Nodejs/server/contracts/*
cp -p build/contracts/* ../Nodejs/server/contracts/
cd ../Nodejs/server
pm2 start myNodeApp.js --log-date-format 'DD-MM HH:mm:ss.SSS'
pm2 logs

![image](https://user-images.githubusercontent.com/25767691/54423733-0cbdd500-471a-11e9-9f75-2a64193e9773.png)

If you delete /build and replace address.json and used the same blockchain datadir
pm2 stop myNodeApp
pm2 delete myNodeApp
cd ../../Solidity/
rm -r build/
truffle deploy --network TestingNetwork
rm ../Nodejs/server/contracts/*
cp -p build/contracts/* ../Nodejs/server/contracts/
cp address.json ../Nodejs/server/helpers/address.json
cd ../Nodejs/server
pm2 start myNodeApp.js --log-date-format 'DD-MM HH:mm:ss.SSS'
pm2 logs

The data is gone (get API) AND the getPastLog API DOESN’T WORK!
![image](https://user-images.githubusercontent.com/25767691/54423765-1ba48780-471a-11e9-88ac-0c7188283f3c.png)

To make getPastLog API work I need to Re-Initiate the blockchain (with a new --datadir) and repeat the process again (where all data is gone)

![image](https://user-images.githubusercontent.com/25767691/54423795-2828e000-471a-11e9-9caa-007989de8449.png)

**Comment by karalabe** [comment]
Geth stores most of the chain state in memory and garbage collects junk historical state in memory. Not writing out everything to disk is essential for reasonable performance.

Of course, we need to push things to disk eventually, so Geth has 3 cases where state gets written to disk: 1) if the cache allowance is reached, older nodes are flushed to disk; 1) if a significant amount of time passed (1h worth of block crunching) to create a snapshot; 3) when you gracefully terminate Geth, anything in memory is flushed to disk.

Now, in your case you did a `kill -9`, which essentially insta-murders Geth, taking everything it held in memory with it to the grave. So the next time you start Geth up, it reverts to the last state that was pushed to disk and syncs from there. If you have a single-node-private-network, there's no place to get the lost data back from.

TL;DR Please make sure to always gracefully terminate Geth so it can push it's hot data out to disk.
